Implemented stieltjes constants

leosartaj opened the pull request:
Implemented `class stieltjes`, that represents, Stieltjes constants,  that occur in Laurent Series expansion of the Riemann zeta function.

```
>>> from sympy import stieltjes
>>> from sympy.abc import n

>>> stieltjes(n)
stieltjes(n)
```

The zero'th constant is the Euler-Mascheroni constant

```
>>> from sympy import stieltjes

>>> stieltjes(0)
EulerGamma
>>> stieltjes(0, 1)
EulerGamma
```

For generalized stieltjes constants

```
>>> from sympy import stieltjes
>>> from sympy.abc import n, m

>>> stieltjes(n, m)
stieltjes(n, m)
```

Constants are only defined for integers >= 0

```
>>> from sympy import stieltjes

>>> stieltjes(-1)
zoo
```

Fixes #4186 

@asmeurer @certik

leosartaj commented:
Please review @asmeurer @certik

certik commented on sympy/functions/special/tests/test_zeta_functions.py:
```diff
@@ -156,3 +156,16 @@ def test_lerchphi_expansion():
+
+    # Zero'th constant EulerGamma
+    assert stieltjes(0) == S.EulerGamma
+    assert stieltjes(0, 1) == S.EulerGamma
```
You should also test `stieltjes(n)`, `stieltjes(n, m)`.

certik commented:
Otherwise I think it looks good.

leosartaj committed:
Implemented stieltjes constants

Represents Stieltjes constants, :math: `\gamma_{k}` that occur in Laurent
Series expansion of the Riemann zeta function.

Examples
========

>>> from sympy import stieltjes
>>> from sympy.abc import n

>>> stieltjes(n)
stieltjes(n)

The zero'th constant is the Euler-Mascheroni constant

>>> from sympy import stieltjes

>>> stieltjes(0)
EulerGamma
>>> stieltjes(0, 1)
EulerGamma

For generalized stieltjes constants

>>> from sympy import stieltjes
>>> from sympy.abc import n, m

>>> stieltjes(n, m)
stieltjes(n, m)

Constants are only defined for integers >= 0

>>> from sympy import stieltjes

>>> stieltjes(-1)
zoo

References
==========

.. [1] http://en.wikipedia.org/wiki/Stieltjes_constants

leosartaj commented:
Added some more tests.

certik commented:
@leosartaj please push new commits in, do not rebase squash commits. It's hard for me to see what was changed since my last review if you just amend your changes into your last commit and force push it in. I think my comment above was still not addressed yet.

leosartaj commented:
Oh, will take care from next time. Also, what kind of test case do you want. An example would help.

certik commented:
What about something like:
```
assert isinstance(stieltjes(n), stieltjes)
```
or something like that. Essentially that things work if you pass in a symbol, as opposed to a number.

debugger22 commented on sympy/functions/special/zeta_functions.py:
```diff
@@ -517,3 +517,69 @@ def eval(cls, s):
+
+
+class stieltjes(Function):
+    """
```
I think this should be a raw string.

leosartaj replied:
I did not get you. What should be a raw string?

debugger22 replied:
You probably don't want `\` for a escape sequence here:
```:math: `\gamma_{k}````

To treat it as a string literal, you need to start the docstring like this:
```r"""your text here"""```

leosartaj replied:
I tried starting the docstring with `r`, but when I remove `\`, I get a `literal` gamma rather than a `symbolic` one.

debugger22 replied:
You don't need to remove `\`.

debugger22 commented on sympy/functions/special/zeta_functions.py:
```diff
@@ -517,3 +517,69 @@ def eval(cls, s):
+
+    >>> from sympy import stieltjes
+    >>> from sympy.abc import n
+
```
Remove this line as it splits the code block generated in the doc. Same for all the following cases.

debugger22 commented on sympy/functions/special/zeta_functions.py:
```diff
@@ -517,3 +517,69 @@ def eval(cls, s):
+    >>> stieltjes(n)
+    stieltjes(n)
+
+    The zero'th constant is the Euler-Mascheroni constant
```
Just zeroth is fine.

debugger22 commented on sympy/functions/special/zeta_functions.py:
```diff
@@ -517,3 +517,69 @@ def eval(cls, s):
+
+    The zero'th constant is the Euler-Mascheroni constant
+
+    >>> from sympy import stieltjes
```
No need of duplicate imports.

leosartaj committed:
added some tests, improved docstring

leosartaj commented:
I have changed the docstring as suggested by @debugger22. Also added some more tests @certik

debugger22 commented:
Looks good to me. +1

leosartaj commented:
Do I need to squash the commits, or is it fine?

certik commented:
This looks good to me, thanks @leosartaj. The failure is unrelated. Merging.
